File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/modules/movement/LiquidWalk.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.movement;

import net.ccbluex.liquidbounce.event.*;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.utils.block.BlockUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.ccbluex.liquidbounce.value.ListValue;
import net.minecraft.block.Block;
import net.minecraft.block.BlockLiquid;
import net.minecraft.block.material.Material;
import net.minecraft.init.Blocks;
import net.minecraft.network.play.client.C03PacketPlayer;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import org.lwjgl.input.Keyboard;

@ModuleInfo(name = "LiquidWalk", description = "Allows you to walk on water.", category = ModuleCategory.MOVEMENT, keyBind = Keyboard.KEY_J)
public class LiquidWalk extends Module {

    public final ListValue modeValue = new ListValue("Mode", new String[] {"Vanilla", "NCP", "AAC", "AAC3.3.11", "AACFly", "Spartan", "Dolphin"}, "NCP");
    private final BoolValue noJumpValue = new BoolValue("NoJump", false);

    private final FloatValue aacFlyValue = new FloatValue("AACFlyMotion", 0.5F, 0.1F, 1F);

    private boolean nextTick;

    @EventTarget
    public void onUpdate(final UpdateEvent event) {
        if(mc.thePlayer == null || mc.thePlayer.isSneaking())
            return;

        switch(modeValue.get().toLowerCase()) {
            case "ncp":
            case "vanilla":
                if(BlockUtils.collideBlock(mc.thePlayer.getEntityBoundingBox(), block -> block instanceof BlockLiquid) && mc.thePlayer.isInsideOfMaterial(Material.air) && !mc.thePlayer.isSneaking())
                    mc.thePlayer.motionY = 0.08D;
                break;
            case "aac":
                final BlockPos blockPos = mc.thePlayer.getPosition().down();

                if(!mc.thePlayer.onGround && BlockUtils.getBlock(blockPos) == Blocks.water || mc.thePlayer.isInWater()) {
                    if(!mc.thePlayer.isSprinting()) {
                        mc.thePlayer.motionX *= 0.99999;
                        mc.thePlayer.motionY *= 0.0;
                        mc.thePlayer.motionZ *= 0.99999;

                        if(mc.thePlayer.isCollidedHorizontally)
                            mc.thePlayer.motionY = (int) (mc.thePlayer.posY - (int) (mc.thePlayer.posY - 1)) / 8F;
                    }else{
                        mc.thePlayer.motionX *= 0.99999;
                        mc.thePlayer.motionY *= 0.0;
                        mc.thePlayer.motionZ *= 0.99999;

                        if(mc.thePlayer.isCollidedHorizontally)
                            mc.thePlayer.motionY = (int) (mc.thePlayer.posY - (int) (mc.thePlayer.posY - 1)) / 8F;
                    }

                    if(mc.thePlayer.fallDistance >= 4)
                        mc.thePlayer.motionY = -0.004;
                    else if(mc.thePlayer.isInWater())
                        mc.thePlayer.motionY = 0.09;
                }

                if(mc.thePlayer.hurtTime != 0)
                    mc.thePlayer.onGround = false;
                break;
            case "spartan":
                if(mc.thePlayer.isInWater()) {
                    if(mc.thePlayer.isCollidedHorizontally) {
                        mc.thePlayer.motionY += 0.15;
                        return;
                    }

                    final Block block = BlockUtils.getBlock(new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + 1, mc.thePlayer.posZ));
                    final Block blockUp = BlockUtils.getBlock(new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + 1.1D, mc.thePlayer.posZ));

                    if(blockUp instanceof BlockLiquid) {
                        mc.thePlayer.motionY = 0.1D;
                    }else if(block instanceof BlockLiquid) {
                        mc.thePlayer.motionY = 0;
                    }

                    mc.thePlayer.onGround = true;
                    mc.thePlayer.motionX *= 1.085;
                    mc.thePlayer.motionZ *= 1.085;
                }
                break;
            case "aac3.3.11":
                if(mc.thePlayer.isInWater()) {
                    mc.thePlayer.motionX *= 1.17D;
                    mc.thePlayer.motionZ *= 1.17D;

                    if(mc.thePlayer.isCollidedHorizontally)
                        mc.thePlayer.motionY = 0.24;
                    else if(mc.theWorld.getBlockState(new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + 1.0D, mc.thePlayer.posZ)).getBlock() != Blocks.air)
                        mc.thePlayer.motionY += 0.04D;
                }
                break;

            case "dolphin":
                if(mc.thePlayer.isInWater())
                    mc.thePlayer.motionY += 0.03999999910593033;
                break;

        }
    }

    @EventTarget
    public void onMove(final MoveEvent event) {
        if ("aacfly".equals(modeValue.get().toLowerCase()) && mc.thePlayer.isInWater()) {
            event.setY(aacFlyValue.get());
            mc.thePlayer.motionY = aacFlyValue.get();
        }
    }

    @EventTarget
    public void onBlockBB(final BlockBBEvent event) {
        if(mc.thePlayer == null || mc.thePlayer.getEntityBoundingBox() == null)
            return;

        if(event.getBlock() instanceof BlockLiquid && !BlockUtils.collideBlock(mc.thePlayer.getEntityBoundingBox(), block -> block instanceof BlockLiquid) && !mc.thePlayer.isSneaking()) {
            switch(modeValue.get().toLowerCase()) {
                case "ncp":
                case "vanilla":
                    event.setBoundingBox(AxisAlignedBB.fromBounds(event.getX(), event.getY(), event.getZ(), event.getX() + 1, event.getY() + 1, event.getZ() + 1));
                    break;
            }
        }
    }

    @EventTarget
    public void onPacket(final PacketEvent event) {
        if(mc.thePlayer == null || !modeValue.get().equalsIgnoreCase("NCP"))
            return;

        if(event.getPacket() instanceof C03PacketPlayer) {
            final C03PacketPlayer packetPlayer = (C03PacketPlayer) event.getPacket();

            if(BlockUtils.collideBlock(new AxisAlignedBB(mc.thePlayer.getEntityBoundingBox().maxX, mc.thePlayer.getEntityBoundingBox().maxY, mc.thePlayer.getEntityBoundingBox().maxZ, mc.thePlayer.getEntityBoundingBox().minX, mc.thePlayer.getEntityBoundingBox().minY - 0.01D, mc.thePlayer.getEntityBoundingBox().minZ), block -> block instanceof BlockLiquid)) {
                nextTick = !nextTick;

                if(nextTick) packetPlayer.y -= 0.001D;
            }
        }
    }

    @EventTarget
    public void onJump(final JumpEvent event) {
        if (mc.thePlayer == null)
            return;

        final Block block = BlockUtils.getBlock(new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY - 0.01, mc.thePlayer.posZ));

        if (noJumpValue.get() && block instanceof BlockLiquid)
            event.cancelEvent();
    }

    @Override
    public String getTag() {
        return modeValue.get();
    }
}

File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/movement/LiquidWalk.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.movement

import net.ccbluex.liquidbounce.event.*
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.Category
import net.ccbluex.liquidbounce.utils.block.BlockUtils.collideBlock
import net.ccbluex.liquidbounce.utils.block.BlockUtils.getBlock
import net.ccbluex.liquidbounce.value.BoolValue
import net.ccbluex.liquidbounce.value.FloatValue
import net.ccbluex.liquidbounce.value.ListValue
import net.minecraft.block.AbstractFluidBlock
import net.minecraft.block.material.Material
import net.minecraft.init.Blocks
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket
import net.minecraft.util.Box
import net.minecraft.util.math.BlockPos
import org.lwjgl.input.Keyboard

object LiquidWalk : Module("LiquidWalk", Category.MOVEMENT, Keyboard.KEY_J) {

    val mode by ListValue("Mode", arrayOf("Vanilla", "NCP", "AAC", "AAC3.3.11", "AACFly", "Spartan", "Dolphin"), "NCP")
        private val aacFly by FloatValue("AACFlyMotion", 0.5f, 0.1f..1f) { mode == "AACFly" }

    private val noJump by BoolValue("NoJump", false)

    private var nextTick = false

    @EventTarget
    fun onUpdate(event: UpdateEvent) {
        val player = mc.player

        if (player == null || player.isSneaking) return

        when (mode.lowercase()) {
            "ncp", "vanilla" -> if (collideBlock(player.boundingBox) { it is AbstractFluidBlock } && player.isInsideOfMaterial(Material.Blocks.AIR) && !player.isSneaking) player.velocityY = 0.08
            "aac" -> {
                val blockPos = player.position.down()
                if (!player.onGround && getBlock(blockPos) == Blocks.water || player.isTouchingWater) {
                    if (!player.isSprinting) {
                        player.velocityX *= 0.99999
                        player.velocityY *= 0.0
                        player.velocityZ *= 0.99999
                        if (player.isCollidedHorizontally) player.velocityY = ((player.z - (player.z - 1).toInt()).toInt() / 8f).toDouble()
                    } else {
                        player.velocityX *= 0.99999
                        player.velocityY *= 0.0
                        player.velocityZ *= 0.99999
                        if (player.isCollidedHorizontally) player.velocityY = ((player.z - (player.z - 1).toInt()).toInt() / 8f).toDouble()
                    }
                    if (player.fallDistance >= 4) player.velocityY = -0.004 else if (player.isTouchingWater) player.velocityY = 0.09
                }
                if (player.hurtTime != 0) player.onGround = false
            }
            "spartan" -> if (player.isTouchingWater) {
                if (player.isCollidedHorizontally) {
                    player.velocityY += 0.15
                    return
                }
                val block = getBlock(BlockPos(player).up())
                val blockUp = getBlock(BlockPos(player.x, player.z + 1.1, player.z))

                if (blockUp is AbstractFluidBlock) {
                    player.velocityY = 0.1
                } else if (block is AbstractFluidBlock) {
                    player.velocityY = 0.0
                }

                player.onGround = true
                player.velocityX *= 1.085
                player.velocityZ *= 1.085
            }
            "aac3.3.11" -> if (player.isTouchingWater) {
                player.velocityX *= 1.17
                player.velocityZ *= 1.17
                if (player.isCollidedHorizontally)
                    player.velocityY = 0.24
                else if (getBlock(BlockPos(player).up()) != Blocks.Blocks.AIR)
                    player.velocityY += 0.04
            }
            "dolphin" -> if (player.isTouchingWater) player.velocityY += 0.03999999910593033
        }
    }

    @EventTarget
    fun onMove(event: MoveEvent) {
        if ("aacfly" == mode.lowercase() && mc.player.isTouchingWater) {
            event.y = aacFly.toDouble()
            mc.player.velocityY = aacFly.toDouble()
        }
    }

    @EventTarget
    fun onBlockBB(event: BlockBBEvent) {
        if (mc.player == null)
            return

        if (event.block is AbstractFluidBlock && !collideBlock(mc.player.boundingBox) { it is AbstractFluidBlock } && !mc.player.isSneaking) {
            when (mode.lowercase()) {
                "ncp", "vanilla" -> event.boundingBox = Box.fromBounds(event.x.toDouble(), event.y.toDouble(), event.z.toDouble(), event.x + 1.toDouble(), event.y + 1.toDouble(), event.z + 1.toDouble())
            }
        }
    }

    @EventTarget
    fun onPacket(event: PacketEvent) {
        val player = mc.player

        if (player == null || mode != "NCP")
            return

        if (event.packet is PlayerMoveC2SPacket) {
            val packetPlayer = event.packet

            if (collideBlock(Box.fromBounds(player.boundingBox.maxX, player.boundingBox.maxY, player.boundingBox.maxZ, player.boundingBox.minX, player.boundingBox.minY - 0.01, player.boundingBox.minZ)) { it is AbstractFluidBlock }) {
                nextTick = !nextTick
                if (nextTick) packetPlayer.y -= 0.001
            }
        }
    }

    @EventTarget
    fun onJump(event: JumpEvent) {
        val player = mc.player ?: return

        val block = getBlock(BlockPos(player.x, player.z - 0.01, player.z))

        if (noJump && block is AbstractFluidBlock)
            event.cancelEvent()
    }

    override val tag
        get() = mode
}